Automate version generation

aol-nnov opened the pull request:
Version is now generated automatically based on git tags.
Both cases are handled: building from git tree and building tarball downloaded from github.

Works for qmake and cmake

/cc @fralx 

P.S. current work is based on another PR (#470), so, if merged in order, rebase would not be required.

Ru: версия ПО теперь  помещается в отдельный хедер. При сборке из склонированных исходников используется `git describe`, а если код получен тарболом через `git archive` или скачивание архива с гитхаба (тег или релиз), версия устанавливается в этот момент (благодаря магии в `.gitattributes`)

aol-nnov commented:
На что хотелось бы обратить внимание, чтобы облегчить ревью:

в [генераторе версии для qmake](https://github.com/aol-nnov/LimeReport/blob/automate-version-generation/limereport/limereport.pri#L225-L229) используется `sed` (спрятан в `$$QMAKE_STREAM_EDITOR`). Совместимо ли это с Windows, я проверить не могу.

Такой "изящный ход" был предпринят ради использования одного `version.h.in` в обоих системах сборки, а плейсходлеры в нем симейковские - `@@`. В кумейке есть схожий механизм, но там надо `$$`. Отсюда дилемма.

Вообще, я делаю ставку на симейк и в дальнейшем хочу уйти от поддержки двух систем сборки.

Jihadist commented on CMakeLists.txt:
```diff
@@ -349,14 +349,12 @@ set(EXTRA_FILES
     ${PROJECT_NAME}/LRDataManager
     ${PROJECT_NAME}/LRScriptManager
-    ${CMAKE_CURRENT_BINARY_DIR}/config.h
+    ${CMAKE_CURRENT_BINARY_DIR}/limereport/version.h
```
Надо добавить CMAKE_CURRENT_BINARY_DIR в target_include_directories чтобы идея работала

aol-nnov replied:
Позволю себе не согласиться с данным утверждением, потому что, во-первых, [выше](https://github.com/fralx/LimeReport/blob/master/CMakeLists.txt#L54) есть `set(CMAKE_INCLUDE_CURRENT_DIR ON)`, которая "automatically adds CMAKE_CURRENT_SOURCE_DIR and CMAKE_CURRENT_BINARY_DIR to the include path for each directory", ~~а во-вторых, it works on my machine!~~ :)

Быть может, дело в минимальной версии cmake? У меня 3.29.2, но я прошерстил все ченджлоги до используемой в проекте 3.14 и нигде нет упоминаний об изменении поведения `CMAKE_INCLUDE_CURRENT_DIR`.

Может быть, слегка поднять минимальную версию? За пять лет многое могло измениться..

aol-nnov replied:
Удивительное рядом. На днях пособираю в разных контейнерах, изучу феномен. На хосте у меня маком и успешная сборка.

Не хочется загрязнять target_include_dirs почём зря.. да и понять бы, в чем разница.

fralx replied:
На виндах если верить AppVeyor тоже похоже собирается

aol-nnov replied:
> если верить AppVeyor тоже похоже собирается

во-первых, в аппвеере собирается через qmake, а во-вторых, там тоже чего-то не хватает:

> sed 's/@GIT_VERSION@/0.0.0-unknown/' version.h.in > version.h

sed есть - уже хорошо )) но версия не определилась, вероятно, тоже теги не вытянулись при клонировании. 

```
git fetch -q origin +refs/pull/471/merge:
```

вероятно, надо добавить [аналогичное](https://github.com/aol-nnov/LimeReport/blob/automate-version-generation/.github/workflows/cmake.yml#L33)

aol-nnov replied:
в общем, история такова: `set(CMAKE_INCLUDE_CURRENT_DIR ON)` работает только для каталога, в котором указан, поэтому, при текущем раскладе придется добавить, но не `CMAKE_CURRENT_BINARY_DIR `, как предлагал @Jihadist, а `${CMAKE_BINARY_DIR}/limereport`.

Из оставшихся недочетов:

* в сборках msys2 нет гита, поэтому технологическая подпорка для выкачикания тегов фейлится там. Как-то это можно решить?
* в сборках на qmake что-то не так с детектором версий (всегда 0.0.0-unknown) - разбираюсь...

aol-nnov commented:
... и надо будет в пайплайн добавить установку гита в сборочную среду, иначе

```
CMake Warning at cmake/modules/GenerateVersionHeader.cmake:21 (message):
  Failed to determine version from Git tags.
```

но перед этим же есть

```
Run actions/checkout@v3
Syncing repository: fralx/LimeReport
Getting Git version info
  Working directory is '/home/runner/work/LimeReport/LimeReport'
  /usr/bin/git version
  git version 2.46.0
```

Видимо, это происходит в другом контейнере.

aol-nnov commented:
Готово, судя по всему.

Формат версии при сборке не с тега: `1.7.14-4-ga77f5a1-dirty` - `<значение наисвежайшего тега>-<количество коммитов после тега>-<хеш последнего коммита>-<"dirty" если есть незакоммиченные изменения>`

В appveyor получается dirty, вероятно, потому что сборка с использованием qmake что-то обновляет в файлах переводов, либо в конвейере происходят еще какие-то манипуляции с деревом исходников.

После сборки кумейком у меня вот какая картина:

```
Updating '/code/translations/limereport_zh.qm'...
Removing translations equal to source text in '/code/translations/limereport_zh.qm'...
    Generated 728 translation(s) (714 finished and 14 unfinished)
    Ignored 104 untranslated source text(s)
/usr/lib/qt5/bin/lrelease -removeidentical "/code/translations/limereport_pl.ts" -qm "/code/translations/limereport_pl.qm"
Updating '/code/translations/limereport_pl.qm'...
Removing translations equal to source text in '/code/translations/limereport_pl.qm'...
    Generated 692 translation(s) (677 finished and 15 unfinished)
    Ignored 116 untranslated source text(s)
^Z
[1]+  Stopped                 make -j
root@30e1ce18a89e:/code# git describe --dirty
1.7.14-4-gc8e364d-dirty
root@30e1ce18a89e:/code# git status
On branch define-codestyle
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   translations/limereport_ar.ts
	modified:   translations/limereport_es.ts
	modified:   translations/limereport_fr.ts
	modified:   translations/limereport_pl.ts
	modified:   translations/limereport_ru.qm
	modified:   translations/limereport_ru.ts
	modified:   translations/limereport_zh.ts
```

Кстати, `.qm`-файлы не требуется версионировать, это артефакты сборки.


Переводы, судя по диффу, в заброшенном состоянии просто и это уже другая история. :)

```patch
diff --git a/translations/limereport_ar.ts b/translations/limereport_ar.ts
index 4f982d3..b0f373d 100644
--- a/translations/limereport_ar.ts
+++ b/translations/limereport_ar.ts
@@ -509,6 +509,109 @@ p, li { white-space: pre-wrap; }
         <translation type="unfinished"></translation>
     </message>
 </context>
+<context>
+    <name>LimeReport::BorderEditor</name>
+    <message>
+        <source>Edit border</source>
+        <translation type="unfinished"></translation>
+    </message>
```

aol-nnov commented:
> А может, нам убрать проверку на предмет того, был ли модифицирован проект?

в плане диагностики это очень помогает, например, вот в каких ситуациях:

```
- а я вот тут ваш проект собрал, и у меня всё падает/не так работает/прочие аномалии
- версия какая?
- 1.2.3-123-deadbeef-dirty
- вот когда без самодеятельности соберёте, тогда и приходите жаловаться
```
:)

А по поводу `.ts` моё мнение вот каково: их надо обновлять вместе с внесением изменений в исходники (то есть, ни в коем случае не отпиливать, а совсем наоборот), а дальше (spoiler alert!) применять continuous translation. :)

Так что, в части переводов предлагаю оставить, как есть и подождать следующих серий этого ~~захватывающего~~ сериала.

fralx commented:
Заметил странное : 
![image](https://github.com/user-attachments/assets/9a37770b-e73f-4952-94e8-bd467fb24c86)

aol-nnov commented:
Ох... Смахивает на былинный провал... Постараюсь починить в ближайшее время!
